notebooks/stitches_dev/R_dev-update-matching.Rmd

```diff
@@ -0,0 +1,501 @@
+                    target_start_yr, target_end_yr,
+                    target_year, target_fx, target_dx, dist_l2),
+            by = groups) -> 
+  y
+
+print(paste("the nearest neighbor is included in matched points:", all(y$min_l2 == y$dist_l2)))
+```
+
+```{r, echo=FALSE}
+# clean up from tests
+rm(matched)
+rm(matched1)
+rm(matched_list)
+rm(groups)
+rm(x)
+rm(y)
+```
+
+
+# Matching - historical period considerations
+
+Matching to neighborhoods of the nearest-neighbor match for each point of target data comes with a few tricky details in the historical period.
+
+
+First, there's a question of whether all of the neighborhood matches in the historical period are 'True' matches.
+
+
+Second, there's the question of whether we want variations in our stitched data from the actual historical data of the target data.
+
+
+## Historical false duplicates
+Do the basic matching of interest with a neighborhood of 0.01 degC (to keep the number of matches low but more than just nearest neighbor):
+
+```{r}
+matched_data <- match_neighborhood(target_data = target_data, archive_data = archive_subset, tol = 0.01,
+                              drop_hist_duplicates = FALSE)
+head(matched_data)
+```
+
+Note that there are two kinds of matches in the historical period:
+
+1. true near matches, eg target 1850 getting matches from  different realizations and or years.
+2. false duplicates: target 1850 gets 1872 data from realization 13 of SSP126 and SSP585. The metadata of these archive values are different, but the actual data values are identical because we just pasted in the same historical data to every Experiment. 
+
+So we write a helper function to drop false duplicates like these in the historical, while keeping any of the true near matches.
+
+
+```{r}
+matched_data <- match_neighborhood(target_data = target_data, archive_data = archive_subset, tol = 0.01,
+                              drop_hist_duplicates = TRUE)
+```
+
+## History identical to the target data
+The above steps allow for variations in the historical period, away from the nearest-neighbor matches in the historical period. Which, for target data from the actual CMIP experiments, means identical history when nearest neighbor matches are in place. 
+
+If we want to only work with stitched realizations that are identical to the historical period of the target data, we would do:
+
+```{r}
+# TODO: Make this a helper function or an input argument to match_neighborhood
+# instead of coding like this. 
+# Also unsure if 2015 is the cutoff year we would want since the data is smoothed. 
+# So 2015 is a 'future' year but the smoothed values still include some history. 
+
+matched_hist <-  match_neighborhood(target_data = filter(target_data, year < 2015),
+                                    archive_data = archive_subset,
+                                    tol = 0)
+matched_fut <-  match_neighborhood(target_data = filter(target_data, year >= 2015),
```

Comment:
I get an error here

```Error: Can't subset elements that don't exist. x Location 1 doesn't exist. ℹ There are only 0 elements. ```

I am wondering if we are using different inputs, if this is something that will get phased out soon there is no need to spend time trying to fix this error. But wanted to give you a heads up in case this was important.

Reply:
I think this is a real error? Based on the phrasing of the error, It seems like it's failing on the `target_data = filter(target_data, year >= 2015)` part and not  within the call to `match_neighborhood`. Is that the case?

My guess is it's something different in the inputs, like you suggest. This markdown expects the `input_data/target_data.csv` that is on this branch

Reply:
Alright so looking into it a bit... I think what is going on here is that `drop_hist_false_duplicates` function which is called from within `match_neighborhood` needs target data from before 2010 but we have filtered it out here.

Reply:
If the `filter(target_data, year >= 2015)` filter call is removed it runs 👍